server/src/test/java/org/elasticsearch/recovery/RelocationIT.java

```diff
@@ -587,17 +590,21 @@ public void testRelocateWhileContinuouslyIndexingAndWaitingForRefresh() {
             .execute();
         logger.info("--> index 100 docs while relocating");
         for (int i = 20; i < 120; i++) {
-            client().prepareIndex("test", "type", Integer.toString(i)).setRefreshPolicy(WriteRequest.RefreshPolicy.WAIT_UNTIL)
-                .setSource("field", "value" + i).execute();
+            pendingIndexResponses.add(client().prepareIndex("test", "type", Integer.toString(i))
+                .setRefreshPolicy(WriteRequest.RefreshPolicy.WAIT_UNTIL)
+                .setSource("field", "value" + i).execute());
         }
         relocationListener.actionGet();
         clusterHealthResponse = client().admin().cluster().prepareHealth().setWaitForEvents(Priority.LANGUID)
             .setWaitForNoRelocatingShards(true).setTimeout(ACCEPTABLE_RELOCATION_TIME).execute().actionGet();
         assertThat(clusterHealthResponse.isTimedOut(), equalTo(false));
 
         logger.info("--> verifying count");
-        client().admin().indices().prepareRefresh().execute().actionGet();
-        assertThat(client().prepareSearch("test").setSize(0).execute().actionGet().getHits().getTotalHits().value, equalTo(120L));
+        assertBusy(() -> {
+            client().admin().indices().prepareRefresh().execute().actionGet();
+            assertThat(client().prepareSearch("test").setSize(0).execute().actionGet().getHits().getTotalHits().value, equalTo(120L));
+            assertTrue(pendingIndexResponses.stream().allMatch(ActionFuture::isDone));
```

Comment:
More of a general comment on these kinds if tests:

I wonder if it makes sense to make this test a little more deterministic by waiting on all the futures first and then running:

```java
client().admin().indices().prepareRefresh().execute().actionGet();
            assertThat(client().prepareSearch("test").setSize(0).execute().actionGet().getHits().getTotalHits().value, equalTo(120L));
```
?

Reply:
we can move the search out, but the refresh and wait for futures need to stay in assertBusy, because we need the refreshes to complete the indexing.